koschei/backend/file_cache.py

```diff
@@ -0,0 +1,210 @@
+        write lock held.
+        """
+        raise NotImplementedError()
+
+    def _p(self, filename):
+        """
+        Returns path to filename in the cache
+        """
+        return os.path.join(self._cachedir, filename)
+
+    def _read_index(self, silent=False):
+        """
+        Reads entries from index file. If the file is invalid or of older
+        version, returns empty dict (which will later on cause cache discard)
+        Expects the index to be read locked already.
+        """
+        index_path = self._p('index.json')
+        entries = {}
+        if os.path.exists(index_path):
+            with open(index_path) as index:
+                try:
+                    index_content = json.load(index)
+                    if index_content['version'] > self.INDEX_VERSION:
+                        raise CacheVersionMismatch(
+                            "Cache index version is newer than current"
+                        )
+                    elif index_content['version'] < self.INDEX_VERSION:
+                        if not silent:
+                            self.log.info("Cache index version is old. "
+                                          "Discarding cache")
+                    else:
+                        entries = index_content['entries']
+                except CacheVersionMismatch as e:
+                    raise
+                except Exception as e:
+                    if not silent:
+                        self.log.warn(
+                            "Cannot parse cache index %s, discarding cache",
+                            e,
+                        )
+        return entries
+
+    def _write_index(self, entries):
+        """
+        Writes entries to the index file.
+        Expects it to be exclusively locked already.
+        """
+        index_path = self._p('index.json')
+        with open(index_path + '.tmp', 'w') as index:
+            json.dump(dict(version=self.INDEX_VERSION, entries=entries),
+                      index, indent=True)
+        # would need fsync (or O_PONIES) to be truly atomic, but it's just a cache
+        os.rename(index_path + '.tmp', index_path)
+
+    def _cleanup_items(self, entries, exclude):
+        """
+        Removes all directories that can be locked and don't have corresponding
+        entries in ready state.
+        Expects index to be exclusively locked. Writes to index file.
+        """
+        dirents = [d for d in os.listdir(self._cachedir)
+                   if os.path.isdir(self._p(d)) and not d == exclude]
+        for dirent in dirents:
+            if entries.get(dirent) != 'ready':
+                with FileLock(self._cachedir, dirent, exclusive=True,
+                              immediate=False) as lock:
+                    if lock.try_lock():
+                        entries.pop(dirent, None)
+                        self._write_index(entries)  # preserve atomicity
```

Comment:
Not when the processes is killed. And resolver processed are first target of oom-killer, so it's not that unusual condition.